File: 1. Bloque/block.py
import hashlib
import time
from datetime import datetime
from pprint import pprint as pp


class Block():
  def __init__(self, index, transactions, previous_hash, difficulty):
    self.index = index
    self.transactions = transactions
    self.transactions_count = len(transactions)
    self.previous_hash = previous_hash
    self.difficulty = difficulty
    self.nonce, self.timestamp, self.found_date = self._proof_of_work()

    self.block = self._block()

  def _calc_hash(self, nonce, timestamp):
    sha = hashlib.sha256()
    sha.update(
      str(self.index).encode() +
      str(nonce).encode() +
      str(timestamp).encode() + 
      str(self.transactions).encode() +
      str(self.transactions_count).encode() +
      str(self.previous_hash).encode()
    )

    return sha.hexdigest()

  def _proof_of_work(self):
    nonce = 0
    timestamp = datetime.now().timestamp()

    while(True):
      hash_value = self._calc_hash(nonce, timestamp)
      if str(hash_value).startswith(self.difficulty):
        self.hash = hash_value

        return [nonce, timestamp, datetime.now().strftime("%B %d, %Y - %H:%M:%S")]
      else:
        nonce += 1

  def _block(self):
    header = { 'prev_hash': self.previous_hash, 'nonce': self.nonce }
    body = { 'total_ttxs': self.transactions_count, 'ttxs': self.transactions }
    return { 'header': header, 'body': body }

  def _to_show(self):
    return { 
      'hash': self.hash, 
      'block': self.block, 
      'index': self.index, 
      'found_date': self.found_date,
      'timestamp': self.timestamp }

  def genesis(transactions, difficulty):
    block_created = Block(0, transactions, '00000000000000000000000000000000000000000000000000000000000000', difficulty)
    return block_created._to_show()

  def new_block(previous_block, transactions, difficulty):
    block_created = Block(previous_block['index'] + 1, transactions, previous_block['hash'], difficulty)
    return block_created._to_show()

File: Tarea 1/block.rb
require 'digest'
require 'pp'

class Block
  attr_reader :data
  attr_reader :prev
  attr_reader :difficulty
  attr_reader :time
  attr_reader :nonce

  def initialize(data, prev, difficulty = '0000')
    @data = data
    @prev = prev
    @difficulty = difficulty
    @nonce, @time = proof_of_work(difficulty)
  end

  def hash
    Digest::SHA256.hexdigest("#{nonce}#{time}#{difficulty}#{prev}#{data}")
  end

  def proof_of_work(diff='00')
    nonce = 0
    time = Time.now.to_i
    loop do
      hash = Digest::SHA256.hexdigest("#{nonce}#{time}#{diff}#{prev}#{data}")
      if hash.start_with?(diff)
        return [nonce, time]
      else
        nonce += 1
      end
    end
  end
end

b0 = Block.new('Genesis', '0000000000000000000000000000000000000000000000000000000000000000')
b1 = Block.new('Leonardo', b0.hash)
b2 = Block.new('Karen', b1.hash)
b3 = Block.new('Hai', b2.hash)


blockchain = [b0, b1, b2, b3]

pp blockchain
# pp b0
# puts '===' * 10
# pp b1
# puts '===' * 10
# pp b2
# puts '===' * 10
# pp b3
# puts '===' * 10